sys/drv/bcm2835_gpio.c

```diff
@@ -71,3 +81,111 @@ void bcm2835_gpio_set_high_detect(resource_t *r, unsigned pin, bool enable) {
+                                       (void **)&pin_cfgs);
+  if (pin_cnt == FDT_PINS_INVAL) {
+    klog("Warning: GPIO FDT entry with no `pins` property");
+    result = EINVAL;
+    goto cleanup;
+  }
+
+  function_cnt = FDT_getencprop_alloc_multi(
+    node, "function", sizeof(*function_cfgs), (void **)&function_cfgs);
+  if (function_cnt == FDT_PINS_INVAL) {
+    klog("Warning: GPIO FDT entry with no `function` property");
+    result = EINVAL;
+    goto cleanup;
+  }
+
+  pull_cnt = FDT_getencprop_alloc_multi(node, "pull", sizeof(*pull_cfgs),
+                                        (void **)&pull_cfgs);
+  if (pull_cnt == FDT_PINS_INVAL) {
+    klog("Warning: GPIO FDT entry with no `pull` property");
+    result = EINVAL;
+    goto cleanup;
+  }
+
+  intr_detect_cnt = FDT_getencprop_alloc_multi(
+    node, "intr_detect", sizeof(*intr_detect_cfgs), (void **)&intr_detect_cfgs);
+  if (intr_detect_cnt == FDT_PINS_INVAL) {
+    klog("Warning: GPIO FDT entry with no `intr_detect` property");
+    result = EINVAL;
+    goto cleanup;
+  }
+
+  for (ssize_t i = 0; i < pin_cnt; i++) {
+    uint32_t pin = pin_cfgs[i];
+
+    /* Take entries one-by-one, until we reach an end. Then repeat the last
+     * entry until all pins are configured.
+     */
+    uint32_t fsel =
+      i < function_cnt ? function_cfgs[i] : function_cfgs[function_cnt - 1];
+    uint32_t pull = i < pull_cnt ? pull_cfgs[i] : pull_cfgs[pull_cnt - 1];
+    uint32_t intr_detect = i < intr_detect_cnt
+                             ? intr_detect_cfgs[i]
+                             : intr_detect_cfgs[intr_detect_cnt - 1];
+
+    assert(fsel <= BCM2835_GPIO_ALT3);
+    assert(pull <= BCM2838_GPIO_GPPUD_PULLDOWN);
+    assert(fsel <= BCM2835_GPIO_ALT3);
+    assert(intr_detect <= 1);
+    bcm2835_gpio_function_select(gpio->gpio, pin, (bcm2835_gpio_func_t)fsel);
+    bcm2835_gpio_set_pull(gpio->gpio, pin, (bcm2838_gpio_gppud_t)pull);
+    bcm2835_gpio_set_high_detect(gpio->gpio, pin, intr_detect);
+  }
+
+cleanup:
+  if (pin_cnt > 0)
+    FDT_free(pin_cfgs);
+  if (function_cnt > 0)
+    FDT_free(function_cfgs);
+  if (pull_cnt > 0)
+    FDT_free(pull_cfgs);
+  if (intr_detect_cnt > 0)
+    FDT_free(intr_detect_cfgs);
+
+  return result;
+}
+
+static int bcm2835_gpio_probe(device_t *dev) {
+  return FDT_is_compatible(dev->node, "brcm,rpi3-gpio");
+}
+
+static int bcm2835_gpio_attach(device_t *dev) {
+  int err;
+  bcm2835_gpio_t *gpio = (bcm2835_gpio_t *)dev->state;
+
+  gpio->gpio = device_take_memory(dev, 0, RF_ACTIVE);
+
+  /* Read configuration from FDT */
+  for (phandle_t node = FDT_child(dev->node); node != FDT_NODEV;
+       node = FDT_peer(node)) {
+    if ((err = bcm2835_gpio_read_fdt_entry(dev, node)))
+      return err;
+  }
+
+  return 0;
+}
+
+driver_t gpio_driver = {
+  .desc = "BCM2835 RPi3 GPIO driver",
+  .size = sizeof(bcm2835_gpio_t),
+  .pass = FIRST_PASS,
+  .probe = bcm2835_gpio_probe,
+  .attach = bcm2835_gpio_attach,
+  .interfaces = {},
```

Comment:
I believe that the convention is to simply omit unused interface array.

Reply:
The dummy `sd_card` driver doesn't omit that field tho. Of course providing this field is not necessary here, so i removed it.